src/ArmTemplates/Creator/TemplateCreators/TemplateCreator.cs

```diff
@@ -1,32 +1,87 @@
-﻿using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Templates.Abstractions;
+﻿using System.Collections.Generic;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Constants;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Templates.Abstractions;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Extractor.Models;
 
 namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Creator.TemplateCreators
 {
-    public class TemplateCreator
+    public abstract class TemplateCreator
```

Comment:
TemplateCreatorBase?

Reply:
Do we need the base suffic? 🤔

Reply:
Made it static. Also changed all child-classes to use static methods. Thanks.

Reply:
>Made it static. Also changed all child-classes to use static methods. 

Why? What was wrong with abstract? 

> Do we need the base suffic? 🤔

Not necessarily, it's just what some people prefer.

Reply:
I did it because of these arguments in my head:
- the class has only static methods.
- it is used not only in current child-classes. Why do we need to keep it an abstract in this case
- there are not fields or other stuff in this "abstract" class, that are used by child classes. So again no reason to leave it as abstract

Reply:
How are you going to mock TemplateCreator when testing any of the classes that use it?

> Also I am not a fan of leaving a lot of methods in inheritance without a reason

If you don't want them to inherit from it, why not pass it as a ctor param instead of making it static?

> It is used in extractors also (creators + extractors both)

Can we pass it to the Extractor as a parameter then as well?

If you don't agree, please ask Tom for his opinion and let's go what what he picks.

Reply:
> How are you going to mock TemplateCreator when testing any of the classes that use it?
I do not want to mock this. I think mocks are needed only for some external services. But this TemplateCreator is basically a business logic of an application. We want to check this specific behaviour.

> If you don't want them to inherit from it, why not pass it as a ctor param instead of making it static?
I am not a fan of creating an instance of a class only for calling it's method. It is the reason static is better here IMO. 

> Can we pass it to the Extractor as a parameter then as well?
I am not a fan of adding a constructor for a lot of objects (you will only use it like this - `new Extractor(new TemplateCreator())`. If there is only a single way to use this class (to call it's methods), why do we need to pass it as a parameter?

In general I am not understanding the reason to leave some classes not static, when they are not having any state. I.e. we have a class Person:
```
class Person
{
   public string Name {get;set;}
}
```

This class has some state or information to hold.
Let's view TemplateCreator:
```
public static class TemplateCreator
{
    public static Template GenerateEmptyPropertyTemplateWithParameters() { ... }

    public static Template GenerateEmptyApiTemplateWithParameters(ExtractorParameters extractorParameters) { ... }

    public static Template GenerateEmptyTemplate() { ... }

    public static Template CreateEmptyParameters() { ... }
}
```

It does not hold any state or any information - it only has methods to generate new Templates.
Basically we can name it like `TemplateExtensions` or smth like this. For me `Creator` in name of class also signifies that it is more likely to be a static class, than an instantiated one.

@tomkerkhove, could you, please, kindly view this thread?

Reply:
I don't really have an opinion on this but I understand both sides. However, I'm personally only using static for small utilities and not main chuncks of business logic.

Mocking is one reason, another is that you can stub them and are easier to test indeed but also extensibility. But again, I don't have a strong opinion on it.

Reply:
> I do not want to mock this. I think mocks are needed only for some external services. But this TemplateCreator is basically a business logic of an application. We want to check this specific behaviour.

Yes, and we would check **once** it because it's be its own class that we would have UTs for. If it's static, you need to test it in every single class where you use it, because you can't mock it. And if you use a static class within a static class that's then used in a class that you want to test, you'll probably have to test things that are only remotely related to the class you're testing.

> I am not a fan of creating an instance of a class only for calling it's method. It is the reason static is better here IMO.

My arguments from above apply here. You can't properly test static things.

> I am not a fan of adding a constructor for a lot of objects (you will only use it like this - new Extractor(new TemplateCreator()). 

Ideally we'd want to have autofac set up. You can then create a TemplateCreator as a single instance because it's stateless and then pass it everywhere as a ctor param.

> If there is only a single way to use this class (to call it's methods), why do we need to pass it as a parameter?

So that you can test it :) I feel like this entire discussion comes down to the fact that you cannot properly test things you can't mock and that you can't mock static calls.

> In general I am not understanding the reason to leave some classes not static, when they are not having any state

Same as above - because you want to be able to test it and mock it. It's the basic concept of dependency injection that you want to make classes independent from its dependencies. Without it, you do a change a single static class and now you have to fix 500 tests that have nothing to do with the static class because other classes that used the static class were not independent from it.

Reply:
Pushed changes, made it using inheritance. Thanks for your opinions and help!)